zcash_client_backend/src/tor/http/exchange_rate.rs

```diff
@@ -0,0 +1,427 @@
+use futures_util::join;
+use hyper::StatusCode;
+use iso_currency::Currency;
+use rand::{seq::IteratorRandom, thread_rng};
+use rust_decimal::Decimal;
+use serde::Deserialize;
+use tracing::{error, trace};
+
+use crate::tor::{Client, Error};
+
+impl Client {
+    /// Fetches the latest `currency`/ZEC exchange rate, derived from several exchanges.
+    ///
+    /// Supported currencies:
+    /// - USD
+    ///
+    /// Returns:
+    /// - `Ok(Some(rate))` if at least one exchange request succeeds.
+    /// - `Ok(None)` if the given currency is unsupported.
+    /// - `Err(_)` if none of the exchange queries succeed.
+    pub async fn get_exchange_rate(&self, currency: Currency) -> Result<Option<Decimal>, Error> {
+        let pair = match ExchangePair::get(currency) {
+            Some(pair) => pair,
+            None => return Ok(None),
+        };
+
+        // Fetch the data in parallel.
+        let res = join!(
+            Binance::query(self, pair),
+            Coinbase::query(self, pair),
+            GateIo::query(self, pair),
+            Gemini::query(self, pair),
+            KuCoin::query(self, pair),
+            Mexc::query(self, pair),
+        );
```

Comment:
Why not have a trait that defines the `query` method, so that users of the library can provide their own currency retrieval? Hardcoding this set of endpoints seems pretty limiting.

Reply:
My goal here was for one new API that was straightforward to use, and then we could expand later as needed. This part has gone through multiple complete redesigns, and as of late Friday evening was at a point where this might be feasible (it only just gained API support for multi-currency, though I could not enable it as almost every exchange only offers `USD(T)/ZEC`).

I don't think a `query` trait is sufficient on its own, because an important part of this API is how the individual query results are combined, and I do not want that to be up to the caller. The caller would instead need a way to provide a heterogeneous set of queryable exchanges over which we can operate, which gets into "implement this trait for all tuple lengths up to 32" territory. I can do that, but not in the currently-available time, and it would make the API more complex.